2020/4/passport/src/main.rs

```diff
@@ -0,0 +1,248 @@
+        let mut issue_year = None;
+        let mut expiration_year = None;
+        let mut height = None;
+        let mut hair_color = None;
+        let mut eye_color = None;
+        let mut passport_id = None;
+        let mut country_id = None;
+        let parts = data.split_whitespace();
+        for p in parts {
+            let mut items = p.split(':');
+            let name = items.next().unwrap();
+            let value = items.next().unwrap();
+            match name {
+                "byr" => {
+                    birth_year = Some(String::from(value));
+                }
+                "iyr" => {
+                    issue_year = Some(String::from(value));
+                }
+                "eyr" => {
+                    expiration_year = Some(String::from(value));
+                }
+                "hgt" => {
+                    height = Some(String::from(value));
+                }
+                "hcl" => {
+                    hair_color = Some(String::from(value));
+                }
+                "ecl" => {
+                    eye_color = Some(String::from(value));
+                }
+                "pid" => {
+                    passport_id = Some(String::from(value));
+                }
+                "cid" => {
+                    country_id = Some(String::from(value));
+                }
+
+                &_ => {}
+            }
+        }
+        Passport {
+            birth_year: birth_year,
+            issue_year: issue_year,
+            expiration_year: expiration_year,
+            height: height,
+            hair_color: hair_color,
+            eye_color: eye_color,
+            passport_id: passport_id,
+            country_id: country_id,
+        }
+    }
+
+    pub fn is_valid(&self) -> bool {
+        match &self.birth_year {
+            Some(byr) => is_valid_year(&byr, 1920, 2002) && match &self.issue_year {
+                Some(iyr) => is_valid_year(&iyr, 2010, 2020) && match &self.expiration_year {
+                    Some(eyr) => is_valid_year(&eyr, 2020, 2030) && match &self.height {
+                        Some(h) => is_valid_height(h) && match &self.hair_color {
+                            Some(hcl) => is_valid_hair_color(hcl) && match &self.eye_color {
+                                Some(ecl) => is_valid_eye_color(ecl) && match &self.passport_id {
+                                    Some(pid) => is_valid_passport_id(pid),
+                                    None => false,
+                                },
+                                None => false,
+                            },
+                            None => false,
+                        },
+                        None => false,
+                    },
+                    None => false,
+                },
+                None => false,
+            },
+            None => false,
+        }
+    }
+}
+
+fn is_valid_year(value: &str, min: usize, max: usize) -> bool {
```

Comment:
If you wanted to do error handling here, you'd return `Result<bool, Error>`. Then in the caller you'd just do `is_valid_year(val, min, max)?` to propagate it.

Plumbing `Result` through your code is a little like plumbing `async` in C#, and `?` is your `await` :).